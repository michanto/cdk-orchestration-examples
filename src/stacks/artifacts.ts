import { AppToken, BUILD_TIME, Log, StackToken } from '@michanto/cdk-orchestration';
import { S3FileResource } from '@michanto/cdk-orchestration/orchestration';
import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Function, Runtime, Code } from 'aws-cdk-lib/aws-lambda';
import { IBucket, Bucket } from 'aws-cdk-lib/aws-s3';
import { md5hash } from 'aws-cdk-lib/core/lib/helpers-internal';
import { PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

const NAMESPACE = '@michanto/cdk-orchestration-examples';

/** Token names for use with AppToken and StackToken . */
const NAMED_BUCKET_TKN = `${NAMESPACE}.NamedBucket`;
const UNNAMED_BUCKET_TKN = `${NAMESPACE}.UnnamedBucket`;
const BUILD_ID_TKN = `${NAMESPACE}.BuildId`;

/**
 * Class that creates the tokens we use in our App.
 */
export class SampleTokens {
  /** Resolves to the name of the Named Bucket. */
  static namedBucketNameToken(scope: Construct) {
    return AppToken.string(scope, NAMED_BUCKET_TKN, { displayHint: 'NmdBkt' });
  }

  /**
   * Separate token creation from token resolution.
   * Useful for cross-stack references without having to pass constructs.
   */
  static resolveNamedBucketNameToken(scope: Construct) {
    AppToken.resolveString(scope, NAMED_BUCKET_TKN, {
      produce: () => {
        let stack = Stack.of(scope);
        return `artifact-bucket-${stack.account}-${stack.region}-${
          md5hash(stack.node.path).slice(0, 8).toLowerCase()
        }`;
      },
    });
  }

  /** Resolves to the name of the unnamed Bucket (name will be generated at deployment time). */
  static unnamedBucketNameToken(scope: Construct) {
    return AppToken.string(scope, UNNAMED_BUCKET_TKN, { displayHint: 'UnnmdBkt' });
  }

  /*
  * Separate token creation from token resolution.
  * Useful for cross-stack references without having to pass constructs.
  */
  static resolveUnnamedBucketNameToken(scope: Construct, bucket: IBucket) {
    AppToken.resolveString(scope, UNNAMED_BUCKET_TKN, {
      produce: () => {
        return bucket.bucketName;
      },
    });
  }

  /**
   * Token for the buildId, which defaults to a timestamp, but is configurable.
   *
   * At Synthesis time, this token is resolved by checking a dictionary attached to the
   * App construct.
   *
   * @param scope Any scope in the app will do, as this is an AppToken.
   * @returns A string that prints as `${Token[BLDID.<token id>]}`
   */
  static buildIdToken(scope: Construct) {
    return AppToken.string(scope, BUILD_ID_TKN, {
      displayHint: 'BLDID',
    });
  }

  /**
   * Resolves the token named "CdkOrche::BuildId"
   * Resolved to any, but CDK will call toString() when writing the value since the token
   * is a string. Note BUILD_TIME is a number.
   * @param scope Any scope in the app will do, as this is an AppToken.
   * @param value
   */
  static resolveBuildIdToken(scope: Construct, value: any = BUILD_TIME) {
    AppToken.resolveAny(scope, BUILD_ID_TKN, { produce: () => value });
  }
}


export class OtherStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactBucketsStack', props?: StackProps) {
    super(scope, id, props);
    let log = Log.of(this);

    log.info('// @michanto/cdk-orchestration AppToken //');
    let buildIdToken = AppToken.string(this, 'build_id', {
      displayHint: 'BLDID',
    });
    log.info(JSON.stringify(buildIdToken));
    log.info(this.toJsonString(buildIdToken));

    // Application scoped named token.
    let bucketToken = AppToken.string(this, 'code_bucket', {
      displayHint: 'Bkt',
    });

    log.info('// @michanto/cdk-orchestration StackToken //');
    let app = this.node.root as App;

    let valinor = new Stack(app, `${this.stackName}-Valinor`, {
      env: {
        account: this.account, region: 'us-east-1',
      },
    });

    let beleriand = new Stack(app, `${this.stackName}-Beleriand`, {
      env: {
        account: this.account, region: 'eu-west-1',
      },
    });
    log.info('// Create Stack token in two different stacks. //');
    for (let stack of [valinor, beleriand]) {
      let keyToken = StackToken.string(stack, 'code_key', {
        displayHint: 'Key',
      });
      let codeBucket = Bucket.fromBucketName(stack, 'CodeBucket', bucketToken);
      new Function(stack, 'MyFunction', {
        runtime: Runtime.NODEJS_16_X,
        code: Code.fromBucket(codeBucket, keyToken),
        handler: 'handler',
      });
      StackToken.resolveString(stack, 'code_key', {
        produce: () => `${
          stack.stackName.toLowerCase()
        }/${
          buildIdToken
        }/code.zip`,
      });
    }
    let unresolvedToken = StackToken.string(this, 'unresolved_token', {
      displayHint: 'Unres',
    });
    log.info(unresolvedToken);
    try {
      log.info(`UnresolvedToken = '${this.resolve(unresolvedToken)}'`);
    } catch {
      log.info(`Could not resolve ${unresolvedToken}`);
    }

    StackToken.resolveString(this, 'unresolved_token', { produce: () => 'Now Resolved' });
    try {
      log.info(`UnresolvedToken = '${this.resolve(unresolvedToken)}'`);
    } catch {
      log.info(`Could not resolve ${unresolvedToken}`);
    }

    // Resolve app tokens.
    AppToken.resolveAny(this, 'build_id', { produce: () => BUILD_TIME });
    // Explicitly named bucket for use cross-region.
    let codeBucketName = `code-bucket-${this.account}`;
    new Bucket(this, 'CodeBucket', {
      bucketName: codeBucketName,
    });
    // Resolve bucket name.
    AppToken.resolveString(this, 'code_bucket', { produce: () => codeBucketName });

    log.info(this.resolve(buildIdToken));
  }
}
/**
 * The artifact bucket will be used to store artifact creation products.
 */
export class ArtifactBucketsStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactBucketsStack', props?: StackProps) {
    super(scope, id, props);

    new Bucket(this, 'NamedBucket', {
      bucketName: SampleTokens.namedBucketNameToken(this),
    });
    SampleTokens.resolveNamedBucketNameToken(this);

    const unnamedBucket = new Bucket(this, 'UnnamedBucket', {
    });
    SampleTokens.resolveUnnamedBucketNameToken(this, unnamedBucket);
  }
}

/**
 * Rather than pass the ArtifactBucket to this class, it can use the token.
 */
export class ArtifactWriterStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactWriter', props?: StackProps) {
    super(scope, id, props);
    let namedBucket = Bucket.fromBucketName(this, 'NamedBucket', SampleTokens.namedBucketNameToken(this));
    let unnamedBucket = Bucket.fromBucketName(this, 'UnnamedBucket', SampleTokens.unnamedBucketNameToken(this));

    new S3FileResource(this, 'Written2NamedBucket', {
      bucket: namedBucket,
      key: 'buildid.json',
      body: {
        buildId: SampleTokens.buildIdToken(this),
      },
      physicalResourceId: PhysicalResourceId.of('buildid.json'),
    });
    new S3FileResource(this, 'Written2UnnamedBucket', {
      bucket: unnamedBucket,
      key: 'buildid.json',
      body: {
        buildId: SampleTokens.buildIdToken(this),
      },
      physicalResourceId: PhysicalResourceId.of('buildid.json'),
    });
  }
}
