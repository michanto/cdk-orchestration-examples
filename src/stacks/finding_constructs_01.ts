import { CfnElementUtilities } from '@michanto/cdk-orchestration';
import { Echo } from '@michanto/cdk-orchestration/transforms';
import { Stack, StackProps } from 'aws-cdk-lib';
import { DefinitionBody, StateMachine } from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';
import { GreetingLambdaTask } from '../constructs/greeting_lambda_task';
import { HitlTestStepFunctionDefinition } from '../constructs/hitl_test_step_fn';
import { PyStepFunctionsImport } from '../constructs/py_step_functions_cleanup';


export class FindingConstructs01 extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const succeedStepFunction = new HitlTestStepFunctionDefinition(this, 'HitlStepFunction', {
      successMode: true,
    });
    new StateMachine(this, 'MockHitlStateMachine', {
      definitionBody: DefinitionBody.fromChainable(succeedStepFunction),
    });
    new GreetingLambdaTask(this, 'GreetingLambdaTask', {
      greeting: 'Hello, everyone!',
    });

    new PyStepFunctionsImport(this, 'PyStepFunctionsImport');

    /** /
    let resources = new CfnElementUtilities().cfnResources(this);
    console.log(`${resources.length} resources.`);
    console.log(resources.map(x => `${x.node.path} (${x.cfnResourceType})`));

    /** /
    let lambdaTask = resources.filter(x => x.cfnResourceType == 'Custom::LambdaTask').pop()!;
    new Echo(lambdaTask, 'Echo');
    /** /
    new RemoveSalt(lambdaTask);

    /** END */
  }
}
