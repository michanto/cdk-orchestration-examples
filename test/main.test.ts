import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { FindingConstructs01 } from '../src/stacks/finding_constructs_01';

test('Snapshot', () => {
  const app = new App();
  const stack = new FindingConstructs01(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});