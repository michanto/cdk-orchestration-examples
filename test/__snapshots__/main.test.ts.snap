// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEOnEventB60C9E76": {
      "DependsOn": [
        "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCERole685CF4F2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e1771e52fcea647ccedead3fbfdccbf0d045bfea362e68f66855e4947530193d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCERole685CF4F2",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEvent4AE15604": {
      "DependsOn": [
        "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleDefaultPolicy9B953B1F",
        "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleCBAF3CDB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3542be390685e0c8353d92ccb5796d343cd93ca946b6b0de798004206a199adc.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (test/CDKORCHCUSTOMRESOURCE::Resources/CDKORCHCUSTOMRESOURCEProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEOnEventB60C9E76",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleCBAF3CDB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleCBAF3CDB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleDefaultPolicy9B953B1F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEOnEventB60C9E76",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEOnEventB60C9E76",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleDefaultPolicy9B953B1F",
        "Roles": [
          {
            "Ref": "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEventServiceRoleCBAF3CDB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCERole685CF4F2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GreetingLambdaTask6113D194": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "GreetingLambdaTaskCustomResourcePolicyDC090EA0",
      ],
      "Properties": {
        "EncodedProperties": {
          "Fn::Base64": {
            "Fn::Join": [
              "",
              [
                "{"Create":{"service":"Lambda","action":"invoke","parameters":{"FunctionName":"",
                {
                  "Ref": "GreetingLambdaTaskReverse66C94AF9",
                },
                "","Payload":"{\\"Greeting\\":\\"Hello, everyone!\\"}"},"physicalResourceId":{"id":"Invoke",
                {
                  "Ref": "GreetingLambdaTaskReverse66C94AF9",
                },
                "-c6ef627ecf342fd2e68e8caeb16b5314"}},"Update":{"service":"Lambda","action":"invoke","parameters":{"FunctionName":"",
                {
                  "Ref": "GreetingLambdaTaskReverse66C94AF9",
                },
                "","Payload":"{\\"Greeting\\":\\"Hello, everyone!\\"}"},"physicalResourceId":{"id":"Invoke",
                {
                  "Ref": "GreetingLambdaTaskReverse66C94AF9",
                },
                "-c6ef627ecf342fd2e68e8caeb16b5314"}},"ResponseBufferField":"Payload","RequestedOutputs":[],"salt":1719520416787}",
              ],
            ],
          },
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCEProviderframeworkonEvent4AE15604",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LambdaTask",
      "UpdateReplacePolicy": "Delete",
    },
    "GreetingLambdaTaskCustomResourcePolicyDC090EA0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:Invoke",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GreetingLambdaTaskCustomResourcePolicyDC090EA0",
        "Roles": [
          {
            "Ref": "CDKORCHCUSTOMRESOURCEResourcesCDKORCHCUSTOMRESOURCERole685CF4F2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GreetingLambdaTaskReverse66C94AF9": {
      "DependsOn": [
        "GreetingLambdaTaskReverseServiceRoleE51874E3",
      ],
      "Properties": {
        "Code": {
          "ZipFile": ""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reverseGreeting = reverseGreeting;
const logger = process.env.LogLevel ? console : undefined;
function log(message) {
logger === null || logger === void 0 ? void 0 : logger.log(JSON.stringify(message));
}

async function reverseGreeting(event, context) {
var _a, _b;
log({ Event: event });
log({ Context: context });
let isCustomResource = event.ResourceProperties != undefined;
if (isCustomResource && 'EncodedProperties' in event.ResourceProperties) {
event.ResourceProperties = {
...event.ResourceProperties,
...JSON.parse(Buffer.from(event.ResourceProperties.EncodedProperties, 'base64').toString('utf8')),
};
log({ DecodedEvent: JSON.stringify(event) });
}
let greeting = (_b = (_a = event.Greeting) !== null && _a !== void 0 ? _a : event.ResourceProperties.Greeting) !== null && _b !== void 0 ? _b : 'Hello, world';
let result = greeting.split('').reverse().join('');
return isCustomResource ? Promise.resolve({
Data: { Greeting: result },
}) : Promise.resolve({
Greeting: result,
});
}",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.reverseGreeting",
        "Role": {
          "Fn::GetAtt": [
            "GreetingLambdaTaskReverseServiceRoleE51874E3",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GreetingLambdaTaskReverseServiceRoleE51874E3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HitlStepFunctionEchoLambda20867270": {
      "DependsOn": [
        "HitlStepFunctionEchoLambdaServiceRoleFB48D8D1",
      ],
      "Properties": {
        "Code": {
          "ZipFile": ""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = handler;
function log(message) {
if (process.env.LogLevel) {
console.log(JSON.stringify(message));
}
}
async function handler(event, context) {
log({ Event: event });
log({ Context: context });
return Promise.resolve(event !== null && event !== void 0 ? event : {});
}",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "LogLevel": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "HitlStepFunctionEchoLambdaServiceRoleFB48D8D1",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "HitlStepFunctionEchoLambdaServiceRoleFB48D8D1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MockHitlStateMachine1D7D3E53": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "MockHitlStateMachineRoleDefaultPolicyB4C117F2",
        "MockHitlStateMachineRole38B152A8",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"AreRunsComplete?","States":{"AreRunsComplete?":{"Type":"Pass","Result":{"s3Uri":"s3://hitl-test-runs/runs/hitl-host-02/2024-05-10T10:38:27Z/15a85c7d-d7ca-43b4-906a-818488f39177","allRunsComplete":true,"allRunsSuccessful":true},"Next":"EchoStep"},"EchoStep":{"Next":"AreRunsSuccessful?","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Resource":"",
              {
                "Fn::GetAtt": [
                  "HitlStepFunctionEchoLambda20867270",
                  "Arn",
                ],
              },
              ""},"AreRunsSuccessful?":{"Type":"Choice","Choices":[{"And":[{"Variable":"$.allRunsComplete","BooleanEquals":true},{"Variable":"$.allRunsSuccessful","BooleanEquals":true}],"Next":"SucceedStep"}],"Default":"FailStep"},"FailStep":{"Type":"Fail","Comment":"All runs complete but not successful","CausePath":"States.JsonToString($)"},"SucceedStep":{"Type":"Succeed","Comment":"All runs complete and successful"}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "MockHitlStateMachineRole38B152A8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "MockHitlStateMachineRole38B152A8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MockHitlStateMachineRoleDefaultPolicyB4C117F2": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "HitlStepFunctionEchoLambda20867270",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "HitlStepFunctionEchoLambda20867270",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MockHitlStateMachineRoleDefaultPolicyB4C117F2",
        "Roles": [
          {
            "Ref": "MockHitlStateMachineRole38B152A8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PyStepFunctionsImportSFRole7AFE3505": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineComponent": {
      "Properties": {
        "DefinitionString": "{
  "StartAt": "MyPassState",
  "States": {
    "MyPassState": {
      "Type": "Pass",
      "Next": "Wait for 3 seconds"
    },
    "Wait for 3 seconds": {
      "Seconds": 3,
      "Type": "Wait",
      "Next": "Convert HelloWorld to Base64"
    },
    "Convert HelloWorld to Base64": {
      "Parameters": {
        "FunctionName": "MyLambda",
        "Payload": {
          "input": "HelloWorld"
        }
      },
      "Resource": "arn:aws:states:::lambda:invoke",
      "Type": "Task",
      "End": true,
      "Retry": [
        {
          "ErrorEquals": [
            "States.TaskFailed"
          ],
          "IntervalSeconds": 15,
          "MaxAttempts": 2,
          "BackoffRate": 4.0
        }
      ],
      "Catch": [
        {
          "ErrorEquals": [
            "States.TaskFailed"
          ],
          "Next": "LambdaTaskFailed"
        }
      ]
    },
    "LambdaTaskFailed": {
      "Type": "Fail"
    }
  }
}",
        "RoleArn": {
          "Fn::GetAtt": [
            "PyStepFunctionsImportSFRole7AFE3505",
            "Arn",
          ],
        },
        "StateMachineName": "MyWorkflow",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
